### PR TITLE
Retire cmp() function

### LINKED ISSUE #493: remove 'from past.builtins import cmp'
The [`cmp`](https://docs.python.org/2/library/functions.html#cmp) function is gone. Importing it from `past` (which is part of the `future` package) is causing deprecation warnings all over the place due to https://github.com/PythonCharmers/python-future/issues/551. The `imp` module [will be removed in Python 3.10](https://www.python.org/dev/peps/pep-0594).
Instances where `cmp` are used should be rewritten and the `cmp` import removed.

This is also -as far as I can tell- the only remaining* reason for the `future` package dependency, which could then potentially be dropped. This would be a good thing as [the `future` package is a bit special](https://github.com/cctbx/dxtbx/blob/2a6b0d34ffde908d2e40072d3de0acfd9ee502c4/libtbx_refresh.py#L12).

(* There are two `range` imports from `past` in `prime`, which I'll address separately.)

### PR DESCRIPTION
**Pull request by Anthchirp**
I rewrote all places where `cmp` is used and removed the `cmp` import.

In the interest of maximum compatibility I did not remove the declarations of the following functions that are now no longer needed, but are declared in public namespaces:
https://github.com/cctbx/cctbx_project/blob/62adfccbbbd622275f8a7166361c644a1cf11511/cctbx/euclidean_model_matching.py#L251-L252
https://github.com/cctbx/cctbx_project/blob/62adfccbbbd622275f8a7166361c644a1cf11511/cctbx/euclidean_model_matching.py#L465-L466
https://github.com/cctbx/cctbx_project/blob/62adfccbbbd622275f8a7166361c644a1cf11511/cctbx/sgtbx/__init__.py#L832-L833
https://github.com/cctbx/cctbx_project/blob/62adfccbbbd622275f8a7166361c644a1cf11511/cctbx/sgtbx/direct_space_asu/check_redundancies.py#L237-L238

#### Background:
The [`cmp`](https://docs.python.org/2/library/functions.html#cmp) function is gone. Importing it from `past` (which is part of the `future` package) is causing deprecation warnings all over the place due to https://github.com/PythonCharmers/python-future/issues/551. The `imp` module [will be removed in Python 3.10](https://www.python.org/dev/peps/pep-0594).

This also -as far as I can tell- would allow retiring the dependency on the `future` package. This is a good thing as [the `future` package is a bit special](https://github.com/cctbx/dxtbx/blob/2a6b0d34ffde908d2e40072d3de0acfd9ee502c4/libtbx_refresh.py#L12).

Closes #493

### CODE REVIEW

**Comment by Anthchirp** [comment]
If there are no concerns I'll merge this on the coming weekend

**Comment by bkpoon** [comment]
I'll have more time next week to look through this, but here are a few points based on a cursory look.

1. There is code duplication where the `cmp` function is implemented several times (`if/else` blocks that return -1, 0, or 1). Please put the [standard implementation of `cmp`](https://python-future.org/_modules/past/builtins/misc.html) that does not use `if/else` in `libtbx/math_utils.py`. The `round2` function is there, so might as well put `cmp` in there too.
2. Currently, the `mmtbx` tests are not being run on Azure Pipelines, so they have to be run manually. I'll run them in a `phenix` build. @olegsobolev can you take a look as well?
3. Some of the `lambda` functions can be replaced with `itemgetter` or `attrgetter` from the `operator` module. This was already done for some other sorting calls that used `lambda` functions.

**Comment by Anthchirp** [comment]
I've made the requested changes.

It would be great to get this in because [the deprecation warnings are making it difficult to spot real issues:](https://dev.azure.com/azure-dials/dials/_build/results?buildId=709&view=results)

image

**Comment by phyy-nx** [comment]
Rebasing onto master should fix the XFEL CI tests.

**Comment by phyy-nx** [comment]
Ah, I see the XFEL CI tests are run twice, once on the branch which is behind master, and once on the merged code (near the bottom of the list) where they are passing.  No worries then.

File: cctbx/euclidean_model_matching.py [hunk 1]
```diff
@@ -249,8 +251,13 @@ def generate_singles(n, i):
...
 def pair_sort_function(pair_a, pair_b):
-  from past.builtins import cmp
-  return cmp(pair_a[0], pair_b[0])
+  # Deprecated. Do not use
+  if pair_a[0] < pair_b[0]:
+    return -1
+  elif pair_a[0] > pair_b[0]:
+    return 1
+  else:
+    return 0
```
Comment (bkpoon) [inline cctbx/euclidean_model_matching.py:260]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/euclidean_model_matching.py [hunk 2]
```diff
@@ -459,10 +465,17 @@ def get_transformed_model2(self,output_pdb=None,
...
+  if len(match_a.pairs) < len(match_b.pairs):
+    return 1
+  elif len(match_a.pairs) > len(match_b.pairs):
+    return -1
+  if match_a.rms < match_b.rms:
+    return -1
+  elif match_a.rms > match_b.rms:
+    return 1
+  else:
+    return 0
```
Comment (bkpoon) [inline cctbx/euclidean_model_matching.py:478]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/sgtbx/__init__.py [hunk 3]
```diff
@@ -829,8 +829,12 @@ def is_equivalent(O, rt_mx_ji):
...
+  # Deprecated. Do not use.
   if (len(a) < len(b)): return -1
   if (len(a) > len(b)): return  1
-  from past.builtins import cmp
-  return cmp(a, b)
+  if a < b:
+    return -1
+  elif a > b:
+    return 1
+  return 0
```
Comment (bkpoon) [inline cctbx/sgtbx/__init__.py:840]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/sgtbx/direct_space_asu/check_redundancies.py [hunk 4]
```diff
@@ -237,12 +235,16 @@ def grid_asu(
...
 
 def compare_redundancies(a, b):
-  return cmp(len(b[1]), len(a[1]))
+  # Deprecated. Do not use
+  if len(b[1]) < len(a[1]):
+    return -1
+  elif len(b[1]) > len(a[1]):
+    return 1
+  else:
+    return 0
```
Comment (bkpoon) [inline cctbx/sgtbx/direct_space_asu/check_redundancies.py:244]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: mmtbx/ions/identify.py [hunk 5]
```diff
@@ -1957,8 +1960,8 @@ def check_ion_environment(self,
                 # XXX probably just O
               self.bad_coords[identity].append(contact)
               inaccuracies.add(self.BAD_COORD_RESIDUE)
-        elif (cmp(0, mmtbx.ions.server.get_charge(contact.atom)) ==
-              cmp(0, ion_params.charge)):
+        elif (_sign(mmtbx.ions.server.get_charge(contact.atom)) ==
+              _sign(ion_params.charge)):
```
Comment (bkpoon) [inline mmtbx/ions/identify.py:1964]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: mmtbx/ions/identify.py [hunk 6]
```diff
@@ -159,6 +158,11 @@ def ion_master_phil():
 WATER, WATER_POOR, LIGHT_ION, HEAVY_ION = range(4)
 NUC_PHOSPHATE_BINDING = ["MG", "CA", "MN"]
 
+def _sign(x):
+  if x > 0: return 1
+  if x < 0: return -1
+  return 0
+
```
Comment (bkpoon) [inline mmtbx/ions/identify.py:165]: Remove since `cmp` is already defined.

File: mmtbx/validation/clashscore.py [hunk 7]
```diff
@@ -9,11 +9,11 @@
 from libtbx.utils import Sorry
 from libtbx import easy_run
 import libtbx.load_env
+import libtbx.math_utils
```
Comment (bkpoon) [inline mmtbx/validation/clashscore.py:12]: Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: mmtbx/validation/clashscore.py [hunk 8]
```diff
@@ -48,7 +48,7 @@ def as_table_row_phenix(self):
              abs(self.overlap) ]
 
   def __cmp__(self, other) : # sort in descending order
-    return cmp(self.overlap, other.overlap)
+    return libtbx.math_utils.cmp(self.overlap, other.overlap)
```
Comment (bkpoon) [inline mmtbx/validation/clashscore.py:51]: Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/array_family/boost_python/tst_flex.py [hunk 9]
```diff
@@ -1113,8 +1112,15 @@ def exercise_operators():
...
-  assert flex.order(b, a) == cmp(tuple(b), tuple(a))
+  if tuple(a) < tuple(b):
+      assert flex.order(a, b) == -1
+      assert flex.order(b, a) == 1
+  elif tuple(a) > tuple(b):
+      assert flex.order(a, b) == 1
+      assert flex.order(b, a) == -1
+  else:
+      assert flex.order(a, b) == 0
+      assert flex.order(b, a) == 0
```
Comment (bkpoon) [inline scitbx/array_family/boost_python/tst_flex.py:1123]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/graph/tardy_tree.py [hunk 10]
```diff
@@ -146,7 +145,9 @@ def cmp_clusters(a, b):
           if (fb): return 1
       if (len(a) > len(b)): return -1
       if (len(a) < len(b)): return 1
-      if (len(a) != 0): return cmp(a[0], b[0])
+      if len(a):
+        if a[0] > b[0]: return 1
+        if a[0] < b[0]: return -1
```
Comment (bkpoon) [inline scitbx/graph/tardy_tree.py:150]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/graph/tardy_tree.py [hunk 11]
```diff
@@ -243,7 +244,9 @@ def cmp_elems(a, b):
       if (fa < fb): return 1
       if (a[1] > b[1]): return -1
       if (a[1] < b[1]): return 1
-      return cmp(a[0], b[0])
+      if (a[0] > b[0]): return 1
+      if (a[0] < b[0]): return -1
+      return 0
```
Comment (bkpoon) [inline scitbx/graph/tardy_tree.py:249]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/euclidean_model_matching.py (hunk 1 above)
Comment (bkpoon) [inline cctbx/euclidean_model_matching.py:260]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/euclidean_model_matching.py (hunk 2 above)
Comment (bkpoon) [inline cctbx/euclidean_model_matching.py:478]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/sgtbx/__init__.py (hunk 3 above)
Comment (bkpoon) [inline cctbx/sgtbx/__init__.py:840]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: cctbx/sgtbx/direct_space_asu/check_redundancies.py (hunk 4 above)
Comment (bkpoon) [inline cctbx/sgtbx/direct_space_asu/check_redundancies.py:244]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: mmtbx/ions/identify.py (hunk 6 above)
Comment (bkpoon) [inline mmtbx/ions/identify.py:165]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: mmtbx/ions/identify.py (hunk 5 above)
Comment (bkpoon) [inline mmtbx/ions/identify.py:1964]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/array_family/boost_python/tst_flex.py (hunk 9 above)
Comment (bkpoon) [inline scitbx/array_family/boost_python/tst_flex.py:1123]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/graph/tardy_tree.py (hunk 10 above)
Comment (bkpoon) [inline scitbx/graph/tardy_tree.py:150]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

File: scitbx/graph/tardy_tree.py (hunk 11 above)
Comment (bkpoon) [inline scitbx/graph/tardy_tree.py:249]: Duplicate `cmp`
Change the import to `from libtbx.math_utils import cmp` so nothing else needs to be changed.

**Comment by bkpoon** [comment]
Looks good, thanks!

**Comment by Anthchirp** [comment]
Thank you for the review!

**Pull request merged**